File: Manipulator3D_Java/src/main/java/manipulator3d/render/DrawUtils.java
package manipulator3d.render;

import static com.raylib.Raylib.*;
import com.raylib.Raylib.*;
import com.raylib.Colors;

/** Rendering helpers for a simple machined-looking manipulator visualization. */
public final class DrawUtils {

    private DrawUtils() {}

    private static void drawTextExAt(Font font, String text, int x, int y, float fontSize, Color color) {
        DrawTextEx(font, text, new Vector2().x(x).y(y), fontSize, 1.0f, color);
    }

    public static void drawTextBold(Font font, String text, int x, int y, float fontSize, Color color) {
        drawTextExAt(font, text, x, y, fontSize, color);
        drawTextExAt(font, text, x + 1, y, fontSize, color);
        drawTextExAt(font, text, x, y + 1, fontSize, color);
        drawTextExAt(font, text, x + 1, y + 1, fontSize, color);
    }

    public static void drawTextSmall(Font font, String text, int x, int y, float fontSize, Color color) {
        drawTextExAt(font, text, x, y, fontSize, color);
    }

    public static void drawRobotBasePedestal(Vector3 origin) {
        // Pedestal below ground slightly + base column
        Vector3 a = new Vector3().x(origin.x()).y(origin.y()).z(-0.25f);
        Vector3 b = new Vector3().x(origin.x()).y(origin.y()).z( 0.00f);
        DrawCylinderEx(a, b, 0.55f, 0.55f, 24, rgba(70,70,75,255));

        Vector3 c = new Vector3().x(origin.x()).y(origin.y()).z(0.00f);
        Vector3 d = new Vector3().x(origin.x()).y(origin.y()).z(0.35f);
        DrawCylinderEx(c, d, 0.38f, 0.34f, 24, rgba(95,95,100,255));

        // Base flange
        DrawCylinderEx(
                new Vector3().x(origin.x()).y(origin.y()).z(0.00f),
                new Vector3().x(origin.x()).y(origin.y()).z(0.06f),
                0.48f, 0.48f, 24,
                rgba(110,110,115,255)
        );
    }

    public static void drawRobotJointHousing(Vector3 center, float radius) {
        DrawSphere(center, radius, rgba(120,120,125,255));
        DrawSphereWires(center, radius, 12, 12, rgba(200,200,200,60));
        DrawCylinderEx(
                new Vector3().x(center.x()).y(center.y()).z(center.z() - 0.10f),
                new Vector3().x(center.x()).y(center.y()).z(center.z() + 0.10f),
                radius * 0.55f, radius * 0.55f, 18,
                rgba(85,85,90,255)
        );
    }

    public static void drawTaperedLink(Vector3 a, Vector3 b, float rA, float rB, Color color) {
        DrawCylinderEx(a, b, rA, rB, 20, color);
        DrawSphere(a, rA * 0.95f, rgba(140,140,145,255));
        DrawSphere(b, rB * 0.95f, rgba(140,140,145,255));
    }

    public static void drawSuctionTool(Vector3 ee, Vector3 approachDir) {
        Vector3 tip = new Vector3()
                .x(ee.x() + approachDir.x() * 0.28f)
                .y(ee.y() + approachDir.y() * 0.28f)
                .z(ee.z() + approachDir.z() * 0.28f);

        DrawCylinderEx(ee, tip, 0.06f, 0.05f, 18, rgba(40,40,45,255));

        Vector3 cupA = new Vector3().x(tip.x()).y(tip.y()).z(tip.z());
        Vector3 cupB = new Vector3()
                .x(tip.x() + approachDir.x() * 0.06f)
                .y(tip.y() + approachDir.y() * 0.06f)
                .z(tip.z() + approachDir.z() * 0.06f);

        DrawCylinderEx(cupA, cupB, 0.11f, 0.11f, 24, rgba(25,25,28,255));
        DrawSphere(tip, 0.035f, rgba(80,80,85,255));
    }

    public static Color rgba(int r, int g, int b, int a) {
        return new Color().r((byte)r).g((byte)g).b((byte)b).a((byte)a);
    }
}

File: Manipulator3D_Ruby/src/render/draw_utils.rb
# frozen_string_literal: true

require_relative "../raylib_bootstrap"
require_relative "../util/math3d"

module Render
  module_function

  def draw_text_ex_at(font, text, x, y, font_size, color)
    DrawTextEx(font, text.to_s, Vector2.create(x.to_f, y.to_f), font_size.to_f, 1.0, color)
  end

  def draw_text_bold(font, text, x, y, font_size, color)
    draw_text_ex_at(font, text, x, y, font_size, color)
    draw_text_ex_at(font, text, x + 1, y, font_size, color)
    draw_text_ex_at(font, text, x, y + 1, font_size, color)
    draw_text_ex_at(font, text, x + 1, y + 1, font_size, color)
  end

  def draw_text_small(font, text, x, y, font_size, color)
    draw_text_ex_at(font, text, x, y, font_size, color)
  end

  def draw_robot_base_pedestal(origin)
    a = Vector3.create(origin.x, origin.y, -0.25)
    b = Vector3.create(origin.x, origin.y, 0.00)
    DrawCylinderEx(a, b, 0.55, 0.55, 24, Color.from_u8(70, 70, 75, 255))

    c = Vector3.create(origin.x, origin.y, 0.00)
    d = Vector3.create(origin.x, origin.y, 0.35)
    DrawCylinderEx(c, d, 0.38, 0.34, 24, Color.from_u8(95, 95, 100, 255))

    DrawCylinderEx(
      Vector3.create(origin.x, origin.y, 0.00),
      Vector3.create(origin.x, origin.y, 0.06),
      0.48, 0.48, 24, Color.from_u8(110, 110, 115, 255)
    )
  end

  def draw_robot_joint_housing(center, radius)
    DrawSphere(center, radius, Color.from_u8(120, 120, 125, 255))
    DrawSphereWires(center, radius, 12, 12, Color.from_u8(200, 200, 200, 60))
    DrawCylinderEx(
      Vector3.create(center.x, center.y, center.z - 0.10),
      Vector3.create(center.x, center.y, center.z + 0.10),
      radius * 0.55, radius * 0.55, 18,
      Color.from_u8(85, 85, 90, 255)
    )
  end

  def draw_tapered_link(a, b, r_a, r_b, color)
    DrawCylinderEx(a, b, r_a, r_b, 20, color)
    DrawSphere(a, r_a * 0.95, Color.from_u8(140, 140, 145, 255))
    DrawSphere(b, r_b * 0.95, Color.from_u8(140, 140, 145, 255))
  end

  def draw_suction_tool(ee, approach_dir)
    tip = Vector3.create(ee.x + approach_dir.x * 0.28, ee.y + approach_dir.y * 0.28, ee.z + approach_dir.z * 0.28)
    DrawCylinderEx(ee, tip, 0.06, 0.05, 18, Color.from_u8(40, 40, 45, 255))

    cup_a = tip
    cup_b = Vector3.create(tip.x + approach_dir.x * 0.06, tip.y + approach_dir.y * 0.06, tip.z + approach_dir.z * 0.06)
    DrawCylinderEx(cup_a, cup_b, 0.11, 0.11, 24, Color.from_u8(25, 25, 28, 255))

    DrawSphere(tip, 0.035, Color.from_u8(80, 80, 85, 255))
  end
end
